dev/io.openliberty.microprofile.faulttolerance.telemetry/src/io/openliberty/microprofile/faulttolerance/telemetry/metrics/integration/AbstractMetricRecorderImpl.java

```diff
@@ -0,0 +1,474 @@
+    private long halfOpenNanos;
+    private long closedNanos;
+    private CircuitBreakerState circuitBreakerState = CircuitBreakerState.CLOSED;
+    protected long lastCircuitBreakerTransitionTime;
+
+    private enum CircuitBreakerState {
+        CLOSED,
+        HALF_OPEN,
+        OPEN
+    }
+
+    //Fields for @Bulkhead Metrics
+    private final LongCounter ftBulkheadCallsTotal;
+    @SuppressWarnings("unused") //This object will be passed into open telemetry by the builder and used there.
+    private final ObservableLongGauge ftBulkheadExecutionsRunning;
+    @SuppressWarnings("unused") //This object will be passed into open telemetry by the builder and used there.
+    private final ObservableLongGauge ftBulkheadExecutionsWaiting;
+    private final LongHistogram ftBulkheadRunningDuration;
+    private final LongHistogram ftBulkheadWaitingDuration;
+
+    private LongSupplier queuePopulationSupplier = null;
+    private LongSupplier concurrentExecutionCountSupplier = null;
+
+    public AbstractMetricRecorderImpl(String classAndMethod, String metricPrefix, Meter meter, RetryPolicy retryPolicy, CircuitBreakerPolicy circuitBreakerPolicy,
+                                      TimeoutPolicy timeoutPolicy,
+                                      BulkheadPolicy bulkheadPolicy, FallbackPolicy fallbackPolicy, AsyncType isAsync) {
+
+        /*
+         * Register all of the metrics required for this method and store them in fields
+         */
+
+        methodAttribute = Attributes.builder().put("method", classAndMethod).build();
+
+        if (retryPolicy != null || timeoutPolicy != null || circuitBreakerPolicy != null || bulkheadPolicy != null || fallbackPolicy != null) {
+            ftInvocationsTotal = meter.counterBuilder(metricPrefix + ".invocations.total").setDescription(Tr.formatMessage(tc, "ft.invocations.total.description")).build();
+            fallbackDefined = (fallbackPolicy != null);
+        } else {
+            ftInvocationsTotal = null;
+            fallbackDefined = false;
+        }
+
+        if (retryPolicy != null) {
+            ftRetryCallsTotal = meter.counterBuilder(metricPrefix + ".retry.calls.total").setDescription(Tr.formatMessage(tc, "ft.retry.calls.total.description")).build();
+            ftRetryRetriesTotal = meter.counterBuilder(metricPrefix + ".retry.retries.total").setDescription(Tr.formatMessage(tc, "ft.retry.retries.total.description")).build();
+        } else {
+            ftRetryCallsTotal = null;
+            ftRetryRetriesTotal = null;
+        }
+
+        if (timeoutPolicy != null) {
+            ftTimeoutCallsTotal = meter.counterBuilder(metricPrefix + ".timeout.calls.total").setDescription(Tr.formatMessage(tc, "ft.timeout.calls.total.description")).build();
+            ftTimeoutExecutionDuration = meter.histogramBuilder(metricPrefix + ".timeout.executionDuration").setDescription(Tr.formatMessage(tc, "ft.timeout.executionDuration.description")).ofLongs().setUnit("nanoseconds").build();
```

Comment:
We should probably have a test for the descriptions - mainly to make sure the NLS strings are substituted correctly.

Reply:
Opened #28480 for this